### PR TITLE
add --no-link-dead-code option to coverage and coveralls

### PR DESCRIPTION
**Pull request by euclio**
Looking for a solution to #43. Not quite sure how to test this. Would appreciate some help!

### CODE REVIEW

**Comment by autohuonw** [comment]
Thanks for the pull request, and welcome! You should hear from @huonw (or someone else) soon.

**Comment by huonw** [comment]
Thanks for the patch! I'm not quite sure this is the right approach: I believe the `--bin` argument only works with the names of actual binaries (i.e. `src/main.rs` or `src/bin/*.rs`), not external tests (i.e. `tests/*.rs`) nor the `#[test]`s in a library. Additionally, I think this should be on by default (I can't think of any reason why someone would want wildly wrong coverage info), but this will need careful handling with respect to compiler versions.

I suspect the best way to handle this would be actually changing cargo to support this, and having `cargo test` (i.e. the `test` profile) pass this option by default.

Thanks again!

**Comment by autohuonw** [comment]
:umbrella: The latest upstream changes (presumably #45) made this pull request unmergeable. Please resolve the merge conflicts.

**Comment by euclio** [comment]
Now that `RUSTFLAGS` has landed in the latest cargo, I figured I'd give this another shot. The flag is now opt-out.

**Comment by euclio** [comment]
Attempted build with this branch here: https://travis-ci.org/Hoverbear/rust-rosetta/builds/117127010#L3956

Failed with "error: multiple input filenames"

**Comment by autohuonw** [comment]
:umbrella: The latest upstream changes (presumably d0cfb8e) made this pull request unmergeable. Please resolve the merge conflicts.

**Comment by huonw** [comment]
Hey @euclio, sorry that I've taken so long. I'm about to give some feedback, but if you no longer have time (or not longer have interest in helping a maintainer as slow as me :) ) I'm happy to take over.

In any case, I've just pushed a little more code that will make this easier:

- it should mean travis CI passes properly
- it adjusts the `run` function to take an `env` parameter for overriding the environment when running the subcommand

Thanks again for getting the ball rolling!

File: travis_cargo.py [hunk 1]
```diff
@@ -173,6 +174,10 @@ def raw_coverage(use_sudo, verify, test_args, merge_msg, kcov_merge_args, kcov_m
     # inconsistent/inaccurate, so using the output of read-manifest is
     # far too much trouble.
 
+    if link_dead_code:
+        flags = ','.join((os.environ.get('RUSTFLAGS', ''), '-C link-dead-code'))
```
Comment (huonw) [inline travis_cargo.py]: Cargo splits `RUSTFLAGS` on spaces, so this should probably be `' '.join(...)`.

File: travis_cargo.py [hunk 2]
```diff
@@ -173,6 +174,10 @@ def raw_coverage(use_sudo, verify, test_args, merge_msg, kcov_merge_args, kcov_m
...
+        os.environ['RUSTFLAGS'] = flags
+
     output = run_output('cargo', 'test', *test_args)
```
Comment (huonw) [inline travis_cargo.py]: This should be adjusted to have an `env = {'RUSTFLAGS': flags}` argument... although I realise the modification I mention only applied to `run`, so `run_output` should probably be adjusted to be similar to the new `run`.

I think we'll want to run `cargo clean` first too: IME, cargo doesn't recompile things when `RUSTFLAGS` changes (maybe a bug in cargo, but travis-cargo has to deal with it).

**Comment by huonw** [comment]
Huh, I realised your approach of using `RUSTFLAGS` is *really neat*: it solves two problems with one stone. It (obviously) gets the `-C link-dead-code` argument through to rustc, *and* it is only interpreted by sufficient new cargo's so the argument can never actually get passed through to a `rustc` that doesn't support that argument (so travis-cargo doesn't have to worry about detecting when the compiler supports it).

Very nice!

**Comment by huonw** [comment]
I didn't hear from you for a while (not as long as you didn't hear from me, but I had some time, so I wanted to make progress!) so I built on your patch to fix it/. I agree that it's tricky to test, so I've just resorted to checking manually, it seems to work: https://coveralls.io/jobs/14197011 (that was all 100% previously).

Thanks again!

**Comment by huonw** [comment]
(Published in 0.1.15, so people will start getting much lower coverage stats now.)

**Comment by euclio** [comment]
Awesome, thanks for getting to this! Glad that it worked out :)

**Pull request merged**